File: fa-vem-s-davem/player/SelectionManager.cs
using System.Collections.Generic;
using System.Linq;

/// Handles selection and deselection of students
public class SelectionManager
{
	private readonly Player owner;
	private readonly List<Selectable> selected = new();

	/// Constructor, sets owner
	public SelectionManager(Player owner)
	{
		this.owner = owner;
	}

	/// Clear selection and select new students
	public void Select(IEnumerable<Selectable> students)
	{
		Clear();

		foreach (var s in students)
		{
			s.Select(owner);
			selected.Add(s);
		}
	}

	/// Clear selection using student.Deselect
	public void Clear()
	{
		foreach (var s in selected.ToList())
			s.Deselect();

		selected.Clear();
	}
	
	/// Append student to selection
	public void Add(IEnumerable<Selectable> students)
	{
		foreach (var s in students)
		{
			if (!selected.Contains(s))
			{
				s.Select(owner);
				selected.Add(s);
			}
		}
	}
	
	/// Remove student from selection
	public void Deselect(Selectable student)
	{
		if (this.selected.Contains(student))
		{
			this.selected.Remove(student);
		}
	}
}

File: src/src/player/SelectionManager.cpp
#include "SelectionManager.hpp"

#include <algorithm>

SelectionManager::SelectionManager(Player *owner) : owner(owner) {}

void SelectionManager::Select(const std::vector<ISelectable *> &students) {
    Clear();

    for (ISelectable *s : students) {
        if (!s)
            continue;

        s->Select(owner);
        selected.push_back(s);
    }
}

void SelectionManager::Clear() {

    // copy because Deselect may modify the vector
    std::vector<ISelectable *> copy = selected;

    for (ISelectable *s : copy) {
        if (s)
            s->Deselect();
    }

    selected.clear();
}

void SelectionManager::Add(const std::vector<ISelectable *> &students) {

    for (ISelectable *s : students) {

        if (!s)
            continue;

        auto it = std::find(selected.begin(), selected.end(), s);

        if (it == selected.end()) {
            s->Select(owner);
            selected.push_back(s);
        }
    }
}

void SelectionManager::Deselect(ISelectable *student) {

    auto it = std::find(selected.begin(), selected.end(), student);

    if (it != selected.end()) {
        selected.erase(it);
    }
}
